deps/rabbit/src/rabbit_variable_queue.erl

```diff
@@ -2613,14 +2614,33 @@ maybe_deltas_to_betas(DelsAndAcksFun,
                         ram_bytes            = RamBytes,
                         disk_read_count      = DiskReadCount,
                         delta_transient_bytes = DeltaTransientBytes,
-                        transient_threshold  = TransientThreshold }) ->
+                        transient_threshold  = TransientThreshold,
+                        rates                = #rates{out = RatesOut }}) ->
     #delta { start_seq_id = DeltaSeqId,
              count        = DeltaCount,
              transient    = Transient,
              end_seq_id   = DeltaSeqIdEnd } = Delta,
+    %% We limit the number of messages we read from the index. We want a large
+    %% enough value but not one that will use too much memory because when there
+    %% are many queues we may end up exploding the memory. We can use the current
+    %% rate to determine the upper value unless configured otherwise. If the rate
+    %% is too low, the min value will be used instead, and it defaults to 100
+    %% as that's neither too low nor too high.
+    %%
+    %% Note that the max ever fetched will be ?SEGMENT_ENTRY_COUNT which currently
+    %% defaults to 16k. So configuring the max to a higher value than that will
+    %% not have an effect (effectively disabling the limit).
+    %%
+    %% We currently only do this for lazy queues because it has not been tested
+    %% with default queues.
+    DeltaConfigEndList = case Mode of
+        lazy -> [DeltaSeqId + max(application:get_env(rabbit, lazy_queue_min_deltas_to_betas, 100),
+                                  application:get_env(rabbit, lazy_queue_max_deltas_to_betas, floor(RatesOut * 1.5)))];
```

Comment:
What was the reasoning behind the multiplication factor, `1.5` ?

Reply:
My guess would be that rates change all the time and we would like to be a little ahead of the delivery rate.

Reply:
Fetch from disk less than once per second. There might be a better default but it would take quite a bit of time to test many different values under different scenarios.

Reply:
Adding details: because the function is used when fetching/prefetching, it makes sense to base it on outgoing rates, and the 1.5 is arbitrary as explained in previous comment.

Reply:
So `floor(RatesOut * 1.5)` stands for "less than once per second"?

I would very much like to see this moved into a function that captures this. Nothing fancy, `less_than_once_per_second(RatesOut)` is sufficient.

Reply:
Happy to make that clearer if there's no objection to the value being what it is.